// sequelize-typescript/dist/browser/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function noop() { return (target, propertyName) => { }; }
class Model {
}
exports.Model = Model;
exports.DataType = {};
exports.AllowNull = noop;
exports.AutoIncrement = noop;
exports.Column = noop;
exports.Comment = noop;
exports.CreatedAt = noop;
exports.Default = noop;
exports.DefaultScope = noop;
exports.DeletedAt = noop;
exports.ForeignKey = noop;
exports.PrimaryKey = noop;
exports.Scopes = noop;
exports.Table = noop;
exports.Unique = noop;
exports.UpdatedAt = noop;
exports.BelongsTo = noop;
exports.BelongsToMany = noop;
exports.HasMany = noop;
exports.HasOne = noop;
exports.AfterBulkCreate = noop;
exports.AfterBulkDelete = noop;
exports.AfterBulkDestroy = noop;
exports.AfterBulkRestore = noop;
exports.AfterBulkSync = noop;
exports.AfterBulkUpdate = noop;
exports.AfterConnect = noop;
exports.AfterCreate = noop;
exports.AfterDefine = noop;
exports.AfterDelete = noop;
exports.AfterDestroy = noop;
exports.AfterFind = noop;
exports.AfterInit = noop;
exports.AfterRestore = noop;
exports.AfterSave = noop;
exports.AfterSync = noop;
exports.AfterUpdate = noop;
exports.AfterUpsert = noop;
exports.AfterValidate = noop;
exports.BeforeBulkCreate = noop;
exports.BeforeBulkDelete = noop;
exports.BeforeBulkDestroy = noop;
exports.BeforeBulkRestore = noop;
exports.BeforeBulkSync = noop;
exports.BeforeBulkUpdate = noop;
exports.BeforeConnect = noop;
exports.BeforeCount = noop;
exports.BeforeCreate = noop;
exports.BeforeDefine = noop;
exports.BeforeDelete = noop;
exports.BeforeDestroy = noop;
exports.BeforeFind = noop;
exports.BeforeFindAfterExpandIncludeAll = noop;
exports.BeforeFindAfterOptions = noop;
exports.BeforeInit = noop;
exports.BeforeRestore = noop;
exports.BeforeSave = noop;
exports.BeforeSync = noop;
exports.BeforeUpdate = noop;
exports.BeforeUpsert = noop;
exports.BeforeValidate = noop;
exports.ValidationFailed = noop;
exports.Contains = noop;
exports.Equals = noop;
exports.Is = noop;
exports.IsAfter = noop;
exports.IsAlpha = noop;
exports.IsAlphanumeric = noop;
exports.IsArray = noop;
exports.IsBefore = noop;
exports.IsCreditCard = noop;
exports.IsDate = noop;
exports.IsDecimal = noop;
exports.IsEmail = noop;
exports.IsFloat = noop;
exports.IsIn = noop;
exports.IsInt = noop;
exports.IsIP = noop;
exports.IsIPv4 = noop;
exports.IsIPv6 = noop;
exports.IsLowercase = noop;
exports.IsNull = noop;
exports.IsNumeric = noop;
exports.IsUppercase = noop;
exports.IsUrl = noop;
exports.IsUUID = noop;
exports.Length = noop;
exports.Max = noop;
exports.Min = noop;
exports.Not = noop;
exports.NotContains = noop;
exports.NotEmpty = noop;
exports.NotIn = noop;
exports.NotNull = noop;
exports.Validate = noop;
exports.Validator = noop;
